src/allocation_engine/interval_tree.rs

```diff
@@ -286,4 +451,158 @@ mod tests {
         ns = NodeState::Free;
         assert!(ns.is_free());
     }
+
+    #[test]
+    fn test_search() {
+        let left_child = InnerNode::new(Range::new(0x90, 0x99).unwrap(), NodeState::Free);
+        let right_child = InnerNode::new(Range::new(0x200, 0x2ff).unwrap(), NodeState::Free);
+        let mut root_node = InnerNode::new(Range::new(0x100, 0x110).unwrap(), NodeState::Allocated);
+        root_node.left = Some(Box::new(left_child.clone()));
+        root_node.right = Some(Box::new(right_child));
+
+        assert_eq!(
+            root_node.search(&Range::new(0x90, 0x99).unwrap()),
+            Some(&left_child)
+        );
+        assert_eq!(root_node.search(&Range::new(0x200, 0x250).unwrap()), None);
+        assert_eq!(root_node.search(&Range::new(0x111, 0x1fe).unwrap()), None);
+    }
+
+    #[test]
+    fn test_search_superset() {
+        let left_child = InnerNode::new(Range::new(0x90, 0x99).unwrap(), NodeState::Free);
+        let right_child = InnerNode::new(Range::new(0x200, 0x2ff).unwrap(), NodeState::Free);
+        let mut root_node = InnerNode::new(Range::new(0x100, 0x110).unwrap(), NodeState::Allocated);
+        root_node.left = Some(Box::new(left_child));
+        root_node.right = Some(Box::new(right_child.clone()));
+
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x100, 0x100).unwrap()),
+            Some(&root_node)
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x200, 0x201).unwrap()),
+            Some(&right_child)
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x200, 0x2ff).unwrap()),
+            Some(&right_child)
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x210, 0x210).unwrap()),
+            Some(&right_child)
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x2ff, 0x2ff).unwrap()),
+            Some(&right_child)
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x2ff, 0x300).unwrap()),
+            None
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x300, 0x300).unwrap()),
+            None
+        );
+        assert_eq!(
+            root_node.search_superset(&Range::new(0x1ff, 0x300).unwrap()),
+            None
+        );
+    }
+
+    #[test]
+    fn test_delete_root_balance_right() {
+        let mut unbalanced_tree =
```

Comment:
I think `delete_root` is defined to be used on balanced_tree.
For example, if you add other larger node (e.g. `right_node5`) on the right subtree, the delete_root will panic because the height diffence during rotation is larger than 2.

Reply:
After discussion with Alex, `test_delete_root_*` are temporary unit tests for `delete_root` function and they will be replaced later when `insert` and `delete` for interval tree are introduced.